python/pynq/intf/fsm_generator.py

```diff
@@ -451,12 +462,16 @@ def parse_fsm_spec(self, fsm_spec_in, use_state_bits):
         self.input_pins = [i[1] for i in fsm_spec['inputs']]
         self.output_pins = [i[1] for i in fsm_spec['outputs']]
 
-        check_num_bits(self.num_input_bits, 'inputs', FSM_MAX_INPUT_BITS)
-        check_num_bits(self.num_output_bits,
-                             'outputs', FSM_MAX_OUTPUT_BITS)
-        check_num_bits(self.num_state_bits, 'states', FSM_MAX_STATE_BITS)
+        check_num_bits(self.num_input_bits, 'inputs',
+                       FSM_MIN_INPUT_BITS, FSM_MAX_INPUT_BITS)
+        check_num_bits(self.num_output_bits, 'outputs',
+                       FSM_MIN_OUTPUT_BITS, FSM_MAX_OUTPUT_BITS)
+        check_num_bits(self.num_state_bits, 'states',
+                       FSM_MIN_STATE_BITS, FSM_MAX_STATE_BITS)
         check_num_bits(self.num_input_bits + self.num_state_bits,
-                             'states and inputs', FSM_MAX_STATE_INPUT_BITS)
+                       'states and inputs',
+                       FSM_MIN_INPUT_BITS + FSM_MIN_STATE_BITS,
+                       FSM_MAX_STATE_INPUT_BITS)
```

Comment:
All these checks are done using constants against class variables: self.* - can we please just do a check_fsm_spec() with no input args and do all these checks in one method - there is alot of docstring bloat to explain each function and also an unnecessary amount of argument passing.